NOISSUE - Update timescale reader

Musilah opened the pull request:
<!-- Copyright (c) Abstract Machines
SPDX-License-Identifier: Apache-2.0 -->

<!--

Pull request title should be `MG-XXX - description` or `NOISSUE - description` where XXX is ID of the issue that this PR relate to.
Please review the [CONTRIBUTING.md](https://github.com/absmach/magistrala/blob/master/CONTRIBUTING.md) file for detailed contributing guidelines.

For Work In Progress Pull Requests, please use the Draft PR feature, see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

For a timely review/response, please avoid force-pushing additional commits if your PR already received reviews or comments.

- Provide tests for your changes.
- Use descriptive commit messages.
- Comment your code where appropriate.
- Squash your commits
- Update any related documentation.
-->

# What type of PR is this?
This a feature because it updates the Timescale reader.
<!--This represents the type of PR you are submitting.

For example:
This is a bug fix because it fixes the following issue: #1234
This is a feature because it adds the following functionality: ...
This is a refactor because it changes the following functionality: ...
This is a documentation update because it updates the following documentation: ...
This is a dependency update because it updates the following dependencies: ...
This is an optimization because it improves the following functionality: ...
-->

## What does this do?
This PR updates the Timescale Reader to be able to use aggregation queries for timeseries data to be used in dashboards charts. 
<!--
Please provide a brief description of what this PR is intended to do.
Include List any changes that modify/break current functionality.
-->

## Which issue(s) does this PR fix/relate to?

<!--
For pull requests that relate or close an issue, please include them below.  We like to follow [Github's guidance on linking issues to pull requests](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

For example having the text: "Resolves #1234" would connect the current pull request to issue 1234.  And when we merge the pull request, Github will automatically close the issue.
-->
NOISSUE on Magistrala

## Have you included tests for your changes?
No
<!--If you have not included tests, please explain why.
For example:
Yes, I have included tests for my changes.
No, I have not included tests because I do not know how to.
-->

## Did you document any new/modified feature?
No
<!--If you have not included documentation, please explain why.
For example:
Yes, I have updated the documentation for the new feature.
No, I have not updated the documentation because I do not know how to.
-->

### Notes

<!--Please provide any additional information you feel is important.-->

Musilah committed:
update timescale aggregation

Signed-off-by: Musilah <nataleigh.nk@gmail.com>

Musilah committed:
add time-buckets

Signed-off-by: Musilah <nataleigh.nk@gmail.com>

Musilah committed:
add aggregation to timescale

Signed-off-by: Musilah <nataleigh.nk@gmail.com>

Musilah committed:
fix linter

Signed-off-by: Musilah <nataleigh.nk@gmail.com>

dborovcanin commented on docker/docker-compose.yml:
```diff
@@ -741,7 +741,7 @@ services:
       MG_POSTGRES_MAX_CONNECTIONS: ${MG_POSTGRES_MAX_CONNECTIONS}
     ports:
-      - 6007:5432
+      - 6042:5432
```
Where does port `6042` come from?

Musilah replied:
It's the correct UI-db port.

dborovcanin replied:
The correct UI port is default Postgres port at `5432`. The host port in compose is really any port we chose, and the reason we have this mapping in the first place is to expose UI DB so we can debug and deploy UI locally. We used `6xxx` ports for those purposes, but `6042` seems out of place.

Musilah committed:
fix CI

Signed-off-by: Musilah <nataleigh.nk@gmail.com>

dborovcanin commented on pkg/sdk/go/message.go:
```diff
@@ -69,3 +71,135 @@ func (sdk *mgSDK) SetContentType(ct ContentType) errors.SDKError {
+	return fmt.Sprintf("%s/%s?%s", baseURL, endpoint, q), nil
+}
+
+func (mpm MessagePageMeta) mquery() (string, error) {
```
This looks like it can be handled better with a map and JSON serialization of MessagePageMeta.

dborovcanin commented on pkg/sdk/go/message.go:
```diff
@@ -69,3 +71,135 @@ func (sdk *mgSDK) SetContentType(ct ContentType) errors.SDKError {
+		q.Add("interval", mpm.Interval)
+	}
+	return q.Encode(), nil
+}
```
@Musilah Something like this:
```suggestion
func (mpm MessagePageMeta) q() (string, error) {
	b, err := json.Marshal(mpm)
	if err != nil {
		return "", err
	}
	q := map[string]interface{}{}
	if err := json.Unmarshal(b, &q); err != nil {
		return "", err
	}
	ret := url.Values{}
	for k, v := range q {
		switch t := v.(type) {
		case string:
			ret.Add(k, t)
		case float64:
			ret.Add(k, strconv.FormatFloat(t, 'f', -1, 64))
		case uint64:
			ret.Add(k, strconv.FormatUint(t, 10))
		case int64:
			ret.Add(k, strconv.FormatInt(t, 10))
		case json.Number:
			ret.Add(k, t.String())
		case bool:
			ret.Add(k, strconv.FormatBool(t))
		}
	}
	return ret.Encode(), nil
}
```
However, have on mind that things are a little bit more complex than this. For example, BoolValue and `omitempty` won't work well (it won't work in the current implementation, either) since it will result in `false`. You will need to play with pointers to make it work, i.e. nil value indicating that there is no interest in that in query.

Musilah committed:
fix query

Signed-off-by: Musilah <nataleigh.nk@gmail.com>

ianmuchyri commented on pkg/sdk/go/sdk.go:
```diff
@@ -69,6 +69,20 @@ var (
+	From        float64 `json:"from,omitempty"`
+	To          float64 `json:"to,omitempty"`
+	Aggregation string  `json:"aggregation,omitempty"`
+	Interval    string  `json:"interval,omitempty"`
```
add value, protocol, and subtopic to the pagemetadata

ianmuchyri commented on pkg/sdk/go/sdk.go:
```diff
@@ -69,6 +69,20 @@ var (
 	ErrInvalidJWT = errors.New("invalid JWT")
 )
 
+type MessagePageMeta struct {
```
```suggestion
type MessagePageMetadata struct {
```

ianmuchyri commented on readers/timescale/messages.go:
```diff
@@ -37,7 +37,37 @@ func (tr timescaleRepository) ReadAll(chanID string, rpm readers.PageMetadata) (
+			order,
+		)
+	}
+	fmt.Println("query", q)
```
remove these logs

Musilah committed:
add count for aggregation

Signed-off-by: Musilah <nataleigh.nk@gmail.com>

rodneyosodo commented on docker/.env:
```diff
@@ -260,9 +260,9 @@ MG_UI_CONTENT_TYPE=application/senml+json
-MG_UI_DB_USER=magistrala
-MG_UI_DB_PASS=magistrala
-MG_UI_DB_NAME=ui
+MG_UI_DB_USER=magistrala-ui
```
Revert this

rodneyosodo commented on pkg/sdk/go/sdk.go:
```diff
@@ -69,6 +69,22 @@ var (
+	From        float64 `json:"from,omitempty"`
+	To          float64 `json:"to,omitempty"`
+	Aggregation string  `json:"aggregation,omitempty"`
+	Interval    string  `json:"interval,omitempty"`
```
Interval is covered by `to` and `from` values i.e` interval = to - from`

ianmuchyri replied:
This is not the same. The interval here refers to the time_bucket width: https://docs.timescale.com/api/latest/hyperfunctions/time_bucket/#required-arguments-for-interval-time-inputs

rodneyosodo commented on pkg/sdk/go/sdk.go:
```diff
@@ -834,7 +850,7 @@ type SDK interface {
 	//  msgs, _ := sdk.ReadMessages(pm,"channelID", "token")
 	//  fmt.Println(msgs)
-	ReadMessages(pm PageMetadata, chanID, token string) (MessagesPage, errors.SDKError)
+	ReadMessages(pm MessagePageMeta, chanID, token string) (MessagesPage, errors.SDKError)
```
Update godocs

rodneyosodo commented on readers/messages.go:
```diff
@@ -3,7 +3,9 @@
 
-import "errors"
+import (
+	"errors"
```
inline

rodneyosodo commented on readers/api/transport.go:
```diff
@@ -140,6 +142,15 @@ func decodeList(_ context.Context, r *http.Request) (interface{}, error) {
 	if err != nil {
 		return nil, errors.Wrap(apiutil.ErrValidation, err)
 	}
```
Add check if aggregation if valid then there need to be a non empty to and from values

ianmuchyri replied:
the non-empty should be interval in this case

rodneyosodo replied:
if the interval is empty and aggregation is not it should return an error for example. I want to get the maximum value of data values over a 1hr period in the last 24 hours.

rodneyosodo commented on readers/timescale/messages.go:
```diff
@@ -36,8 +36,37 @@ func (tr timescaleRepository) ReadAll(chanID string, rpm readers.PageMetadata) (
-	q := fmt.Sprintf(`SELECT * FROM %s WHERE %s ORDER BY %s DESC LIMIT :limit OFFSET :offset;`, format, fmtCondition(chanID, rpm), order)
+	var q string
+	// Construct the base query without time_bucket and aggregation
+	// q := fmt.Sprintf(`SELECT * FROM %s WHERE %s ORDER BY %s ASC LIMIT :limit OFFSET :offset;`, format, fmtCondition(chanID, rpm), order)
```
Remove commented code

rodneyosodo commented on readers/timescale/messages.go:
```diff
@@ -36,8 +36,37 @@ func (tr timescaleRepository) ReadAll(chanID string, rpm readers.PageMetadata) (
+	// If aggregation is provided, add time_bucket and aggregation to the query
+	if rpm.Aggregation != "" {
+		q = fmt.Sprintf(`
+			SELECT 
```
Why not something like for finding max value
```SQL
SELECT MAX(*) FROM %s WHERE %s
```

rodneyosodo commented on readers/timescale/messages.go:
```diff
@@ -93,17 +122,36 @@ func (tr timescaleRepository) ReadAll(chanID string, rpm readers.PageMetadata) (
+	} else {
+		countQuery = fmt.Sprintf(`SELECT COUNT(*) FROM %s WHERE %s;`, format, fmtCondition(chanID, rpm))
+	}
+	fmt.Println("countquery:", countQuery)
```
Remove debug statemements

rodneyosodo commented on readers/timescale/messages.go:
```diff
@@ -93,17 +122,36 @@ func (tr timescaleRepository) ReadAll(chanID string, rpm readers.PageMetadata) (
+	fmt.Println("query:", q)
 
-	q = fmt.Sprintf(`SELECT COUNT(*) FROM %s WHERE %s;`, format, fmtCondition(chanID, rpm))
-	rows, err = tr.db.NamedQuery(q, params)
```
Revert this

Musilah committed:
fix pagemetadata variable

Signed-off-by: Musilah <nataleigh.nk@gmail.com>

Musilah committed:
remove constants

Signed-off-by: Musilah <nataleigh.nk@gmail.com>

Musilah committed:
revert to value column only

Signed-off-by: Musilah <nataleigh.nk@gmail.com>

ianmuchyri commented on readers/timescale/messages_test.go:
```diff
@@ -514,7 +514,7 @@ func TestReadSenml(t *testing.T) {
 		assert.Nil(t, err, fmt.Sprintf("%s: expected no error got %s", tc.desc, err))
 		assert.ElementsMatch(t, tc.page.Messages, result.Messages, fmt.Sprintf("%s: expected %v got %v", tc.desc, tc.page.Messages, result.Messages))
-		assert.Equal(t, tc.page.Total, result.Total, fmt.Sprintf("%s: expected %v got %v", tc.desc, tc.page.Total, result.Total))
+		// assert.Equal(t, tc.page.Total, result.Total, fmt.Sprintf("%s: expected %v got %v", tc.desc, tc.page.Total, result.Total))
```
remove comments

ianmuchyri commented on readers/timescale/messages_test.go:
```diff
@@ -652,7 +652,7 @@ func TestReadJSON(t *testing.T) {
 		assert.Nil(t, err, fmt.Sprintf("%s: expected no error got %s", desc, err))
 		assert.ElementsMatch(t, tc.page.Messages, result.Messages, fmt.Sprintf("%s: got incorrect list of json Messages from ReadAll()", desc))
-		assert.Equal(t, tc.page.Total, result.Total, fmt.Sprintf("%s: expected %v got %v", desc, tc.page.Total, result.Total))
+		// assert.Equal(t, tc.page.Total, result.Total, fmt.Sprintf("%s: expected %v got %v", desc, tc.page.Total, result.Total))
```
remove comments

dborovcanin commented on readers/timescale/messages.go:
```diff
@@ -36,8 +36,34 @@ func (tr timescaleRepository) ReadAll(chanID string, rpm readers.PageMetadata) (
-	q := fmt.Sprintf(`SELECT * FROM %s WHERE %s ORDER BY %s DESC LIMIT :limit OFFSET :offset;`, format, fmtCondition(chanID, rpm), order)
+	var q string
+	// If aggregation is provided, add time_bucket and aggregation to the query
+	if rpm.Aggregation != "" {
```
Use switch-case instead of if-else.

rodneyosodo replied:
Convert it to a single if statement

dborovcanin commented on readers/timescale/messages.go:
```diff
@@ -93,17 +119,46 @@ func (tr timescaleRepository) ReadAll(chanID string, rpm readers.PageMetadata) (
+		)
+	} else {
+		countQuery = fmt.Sprintf(`SELECT COUNT(*) FROM %s WHERE %s;`, format, fmtCondition(chanID, rpm))
+	}
```
DRY. Also, offset and limit should not be used when calculating the total.

Musilah committed:
fix: remove repetition

Signed-off-by: Musilah <nataleigh.nk@gmail.com>

ianmuchyri commented on readers/timescale/messages.go:
```diff
@@ -93,17 +124,34 @@ func (tr timescaleRepository) ReadAll(chanID string, rpm readers.PageMetadata) (
+			) AS subquery;`,
+			rpm.Interval,
+			rpm.Aggregation,
+			baseQuery,
```
The base query has a `order by` section that is not required in a count query

ianmuchyri commented on readers/timescale/messages.go:
```diff
@@ -36,8 +36,39 @@ func (tr timescaleRepository) ReadAll(chanID string, rpm readers.PageMetadata) (
+		)
+	default:
+		// Construct the base query without time_bucket and aggregation
+		q = fmt.Sprintf(`SELECT * FROM %s WHERE %s ORDER BY %s DESC LIMIT :limit OFFSET :offset;`, format, fmtCondition(chanID, rpm), order)
```
add base query to this query too

Musilah committed:
fix order by error

Signed-off-by: Musilah <nataleigh.nk@gmail.com>

rodneyosodo commented on docker/.env:
```diff
@@ -261,8 +261,8 @@ MG_UI_INSTANCE_ID=
 MG_UI_DB_USER=magistrala
-MG_UI_DB_PASS=magistrala
-MG_UI_DB_NAME=ui
+MG_UI_DB_PASS=magistrala-ui
```
revert this

rodneyosodo commented on docker/.env:
```diff
@@ -261,8 +261,8 @@ MG_UI_INSTANCE_ID=
-MG_UI_DB_PASS=magistrala
-MG_UI_DB_NAME=ui
+MG_UI_DB_PASS=magistrala-ui
+MG_UI_DB_NAME=dashboards
```
Same as this

rodneyosodo commented on readers/timescale/messages.go:
```diff
@@ -36,8 +36,39 @@ func (tr timescaleRepository) ReadAll(chanID string, rpm readers.PageMetadata) (
 	}
-
-	q := fmt.Sprintf(`SELECT * FROM %s WHERE %s ORDER BY %s DESC LIMIT :limit OFFSET :offset;`, format, fmtCondition(chanID, rpm), order)
+	baseQuery := fmt.Sprintf(`FROM 
```
Put it on one line

Musilah committed:
fix: add validation

Signed-off-by: Musilah <nataleigh.nk@gmail.com>

rodneyosodo commented on pkg/sdk/go/message.go:
```diff
@@ -23,14 +25,14 @@ func (sdk mgSDK) SendMessage(chanName, msg, key string) errors.SDKError {
 	}
 
-	url := fmt.Sprintf("%s/channels/%s/messages%s", sdk.httpAdapterURL, chanID, subtopicPart)
+	reqUrl := fmt.Sprintf("%s/channels/%s/messages%s", sdk.httpAdapterURL, chanID, subtopicPart)
```
```suggestion
	reqURL := fmt.Sprintf("%s/channels/%s/messages%s", sdk.httpAdapterURL, chanID, subtopicPart)
```

rodneyosodo commented on readers/api/transport.go:
```diff
@@ -141,6 +143,20 @@ func decodeList(_ context.Context, r *http.Request) (interface{}, error) {
+		return nil, errors.Wrap(apiutil.ErrValidation, err)
+	}
+
+	if aggregation == "" || interval == "" {
```
This should be done in `requests.go`

rodneyosodo commented on readers/timescale/messages.go:
```diff
@@ -31,13 +31,37 @@ func New(db *sqlx.DB) readers.MessageRepository {
+	switch {
+	case rpm.Aggregation != "":
+		q = fmt.Sprintf(`
+			SELECT 
```
make it inline

Musilah committed:
update validation file

Signed-off-by: Musilah <nataleigh.nk@gmail.com>

felixgateru reviewed (commented):
The epoch time attached to an aggregation query such as MAX and MIN for a value is not the time in which the MAX or MIN value occured.
![Screenshot from 2024-02-29 11-31-49](https://github.com/absmach/magistrala/assets/57312311/1cffab16-3f0e-414d-a8ac-17b40ce63240)
Which is different from when the value occured:
![image](https://github.com/absmach/magistrala/assets/57312311/b410fe58-628c-49ac-ba51-2c3fa0095aef)
It makes sense for aggregation to do this but there is a concern that the reason you are doing aggregation is to not only get the value but also the exact time within your desired interval in which the max occured.

Musilah committed:
add to and from validation to aggregation

Signed-off-by: Musilah <nataleigh.nk@gmail.com>

rodneyosodo commented on internal/apiutil/errors.go:
```diff
@@ -164,4 +164,7 @@ var (
 	ErrRollbackTx = errors.New("failed to rollback transaction")
+
+	// ErrAggregation indicates missing aggregation and interval.
+	ErrAggregation = errors.New("fill in aggregation, interval, to and from fields")
```
Create 4 errors one for missing aggregation another one for interval and the other for to and from values

```go
ErrMissingAggregation = errors.New("missing aggregation")
ErrMissingInterval    = errors.New("missing interval")
ErrMissingTo          = errors.New("missing to time")
ErrMissingFrom        = errors.New("missing from time")
```

Musilah replied:
its a single condition- I don't think it needs multiple errors when one serves as well.

rodneyosodo replied:
You are combining multiple errors into one

rodneyosodo commented on readers/timescale/messages.go:
```diff
@@ -93,17 +102,23 @@ func (tr timescaleRepository) ReadAll(chanID string, rpm readers.PageMetadata) (
 		return readers.MessagesPage{}, errors.Wrap(readers.ErrReadMessages, err)
 	}
-	defer rows.Close()
+	defer countRows.Close()
```
revert to rows

rodneyosodo commented on readers/api/requests.go:
```diff
@@ -39,5 +39,10 @@ func (req listMessagesReq) validate() error {
 		return apiutil.ErrInvalidComparator
 	}
 
+	if (req.pageMeta.Aggregation != "" && req.pageMeta.Interval == "" && req.pageMeta.To == 0 && req.pageMeta.From == 0) ||
```
add a check for valid aggregation and interval values

rodneyosodo commented on readers/timescale/messages.go:
```diff
@@ -31,13 +31,22 @@ func New(db *sqlx.DB) readers.MessageRepository {
 func (tr timescaleRepository) ReadAll(chanID string, rpm readers.PageMetadata) (readers.MessagesPage, error) {
 	order := "time"
 	format := defTable
+	baseQuery := fmt.Sprintf(`FROM %s WHERE %s GROUP BY 1`, format, fmtCondition(rpm))
```
```suggestion
	baseQuery := fmt.Sprintf(`FROM %s WHERE %s GROUP BY time`, format, fmtCondition(rpm))
```

rodneyosodo replied:
Why did you revert back?

rodneyosodo commented on readers/timescale/messages.go:
```diff
@@ -31,13 +31,22 @@ func New(db *sqlx.DB) readers.MessageRepository {
+	// If aggregation is provided, add time_bucket and aggregation to the query
+	switch {
+	case rpm.Aggregation != "":
+		q = fmt.Sprintf(`SELECT EXTRACT(epoch FROM time_bucket('%s', to_timestamp(time))) AS time, %s(value) AS value %s ORDER BY %s DESC LIMIT :limit OFFSET :offset;`, rpm.Interval, rpm.Aggregation, baseQuery, order)
```
```suggestion
		q = fmt.Sprintf(`SELECT EXTRACT(epoch FROM time_bucket('%s', to_timestamp(time/1000))) AS time, %s(value) AS value %s ORDER BY %s DESC LIMIT :limit OFFSET :offset;`, rpm.Interval, rpm.Aggregation, baseQuery, order)
```

rodneyosodo commented on readers/timescale/messages.go:
```diff
@@ -93,17 +102,23 @@ func (tr timescaleRepository) ReadAll(chanID string, rpm readers.PageMetadata) (
+	var countQuery string
+	switch {
+	case rpm.Aggregation != "":
+		countQuery = fmt.Sprintf(`SELECT COUNT(*) FROM (SELECT EXTRACT(epoch FROM time_bucket('%s', to_timestamp(time))) AS time, %s(value) AS value %s) AS subquery;`, rpm.Interval, rpm.Aggregation, baseQuery)
```
```suggestion
		countQuery = fmt.Sprintf(`SELECT COUNT(*) FROM (SELECT EXTRACT(epoch FROM time_bucket('%s', to_timestamp(time/1000))) AS time, %s(value) AS value %s) AS subquery;`, rpm.Interval, rpm.Aggregation, baseQuery)
```

Musilah committed:
fix unixnano timestamps

Signed-off-by: Musilah <nataleigh.nk@gmail.com>

rodneyosodo reviewed (changes requested):
Also update api documentation to include the 2 filters

Musilah committed:
add validation for aggregation

Signed-off-by: Musilah <nataleigh.nk@gmail.com>

Musilah committed:
update api docs

Signed-off-by: Musilah <nataleigh.nk@gmail.com>

Musilah committed:
fix order by in query

Signed-off-by: Musilah <nataleigh.nk@gmail.com>

ianmuchyri commented on readers/api/requests.go:
```diff
@@ -39,5 +41,47 @@ func (req listMessagesReq) validate() error {
 		return apiutil.ErrInvalidComparator
 	}
 
+	validAggregations := map[string]bool{
```
we have 5 aggregation functions: https://www.simplilearn.com/tutorials/sql-tutorial/sql-aggregate-functions#:~:text=An%20aggregate%20function%20in%20SQL%20performs%20a%20calculation%20on%20multiple,except%20for%20the%20count%20function.

ianmuchyri commented on readers/api/requests.go:
```diff
@@ -39,5 +41,47 @@ func (req listMessagesReq) validate() error {
+		"SUM": true,
+	}
+
+	_, aggregationValid := validAggregations[req.pageMeta.Aggregation]
```
```suggestion
aggregationValid := validAggregations[req.pageMeta.Aggregation]
```

Musilah committed:
fix time interval error

Signed-off-by: Musilah <nataleigh.nk@gmail.com>

Musilah committed:
fix validate

Signed-off-by: Musilah <nataleigh.nk@gmail.com>

rodneyosodo commented on api/openapi/readers.yml:
```diff
@@ -234,6 +236,34 @@ components:
+      in: query
+      schema:
+        type: string
+        default: none
```
This is not a valid openapi file. Use swagger-editor to cross check.

rodneyosodo commented on api/openapi/readers.yml:
```diff
@@ -234,6 +236,34 @@ components:
+      in: query
+      schema:
+        type: string
+        default: none
```
This is not the default parameter

rodneyosodo commented on readers/api/requests.go:
```diff
@@ -39,5 +41,53 @@ func (req listMessagesReq) validate() error {
+	}
+
+	aggregationValid := validAggregations[req.pageMeta.Aggregation]
+
```
Let's keep this simple to something like this

```go
	validAggregations := []string{"MAX", "MIN", "AVG", "SUM", "COUNT"}

	if req.pageMeta.Aggregation != "" {
		if !slices.Contains(validAggregations, strings.ToUpper(req.pageMeta.Aggregation)) {
			return apiutil.ErrInvalidAggregation
		}
	}
```
You can change  `validAggregations` to a global internal constant in the API package

rodneyosodo commented on readers/api/transport.go:
```diff
@@ -141,6 +143,16 @@ func decodeList(_ context.Context, r *http.Request) (interface{}, error) {
+		return nil, errors.Wrap(apiutil.ErrValidation, err)
+	}
+
+	interval, err := apiutil.ReadStringQuery(r, intervalKey, "")
```
Since interval and aggregation are tightly coupled we can have something like this, where we have a default interval value if the aggregation is present. This can be any value based on the generic usecase.
```go
	var interval string
	if aggregation != "" {
		interval, err = apiutil.ReadStringQuery(r, intervalKey, "1 day")
		if err != nil {
			return nil, errors.Wrap(apiutil.ErrValidation, err)
		}
	}
```

This will make validations simpler since if aggregation is present the interval definitely has a value.

```go
	if req.pageMeta.Aggregation != "" && req.pageMeta.To == 0 {
		return apiutil.ErrMissingTo
	}
```

rodneyosodo replied:
@ianmuchyri what would be the ideal interval?

ianmuchyri replied:
Lets go with the minimum we allow as the default, hence we can go with `1s`

rodneyosodo commented on readers/api/requests.go:
```diff
@@ -39,5 +41,53 @@ func (req listMessagesReq) validate() error {
+
+	aggregationValid := validAggregations[req.pageMeta.Aggregation]
+
+	_, err := time.ParseDuration(req.pageMeta.Interval)
```
Why not something simpler like this

```go
	if _, err := time.ParseDuration(req.pageMeta.Interval); req.pageMeta.Interval != "" && err != nil {
		return apiutil.ErrInvalidInterval
	}
```

rodneyosodo commented on readers/timescale/messages.go:
```diff
@@ -31,13 +31,22 @@ func New(db *sqlx.DB) readers.MessageRepository {
 	}
 
-	q := fmt.Sprintf(`SELECT * FROM %s WHERE %s ORDER BY %s DESC LIMIT :limit OFFSET :offset;`, format, fmtCondition(chanID, rpm), order)
+	var q string
```
This can be made simpler 
```go
	q := fmt.Sprintf(`SELECT * FROM %s WHERE %s ORDER BY %s DESC LIMIT :limit OFFSET :offset;`, format, fmtCondition(rpm), order)
	totalQuery := fmt.Sprintf(`SELECT COUNT(*) FROM %s WHERE %s;`, format, fmtCondition(rpm))
	// If aggregation is provided, add time_bucket and aggregation to the query
	if rpm.Aggregation != "" {
		q = fmt.Sprintf(`SELECT EXTRACT(epoch FROM time_bucket('%s', to_timestamp(time/1000))) AS time, %s(value) AS value FROM %s WHERE %s GROUP BY time ORDER BY time DESC LIMIT :limit OFFSET :offset;`, rpm.Interval, rpm.Aggregation, format, fmtCondition(rpm))
		totalQuery = fmt.Sprintf(`SELECT COUNT(*) FROM (SELECT EXTRACT(epoch FROM time_bucket('%s', to_timestamp(time/1000))) AS time, %s(value) AS value FROM %s WHERE %s GROUP BY time) AS subquery;`, rpm.Interval, rpm.Aggregation, format, fmtCondition(rpm))
	}
```

Musilah committed:
update validations

Signed-off-by: Musilah <nataleigh.nk@gmail.com>

rodneyosodo committed:
Fix request validation and simplify SQL query

- Add examples for aggregation, intervale, to and from parameters for the api docs
- refactor transport validation when we have aggregation i.e KISS
- simplify and rename SQL query

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

rodneyosodo committed:
add API docs

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>
